File: src/L03/Fuunctor.scala
package L03

import L01._
import L02._

trait Fuunctor[F[_]] {
  def fmaap[A, B](f: A => B): F[A] => F[B]

  // Provided as a synonym for `fmaap` but with the arguments flipped.
  // This helps the type-inferencer.
  final def apply[A, B](a: F[A])(f: A => B): F[B] =
    fmaap(f)(a)
}

object Fuunctor {
  // Exercise 1
  // Relative Difficulty: 1
  implicit val IdFuunctor: Fuunctor[Id] =
    sys.error("todo")

  // Exercise 2
  // Relative Difficulty: 2
  implicit val ListFuunctor: Fuunctor[List] =
    sys.error("todo")

  // Exercise 3
  // Relative Difficulty: 2
  implicit val OptionalFuunctor: Fuunctor[Optional] =
    sys.error("todo")

  // Exercise 4
  // Relative Difficulty: 3
  implicit def Function1Fuunctor[T]: Fuunctor[({type l[a] = T => a})#l] =
    sys.error("todo")
  
  ///////////////////////
  // SUPPORT LIBRARIES //
  ///////////////////////

  implicit val ScalaListFuunctor: Fuunctor[scala.List] =
    new Fuunctor[scala.List] {
      def fmaap[A, B](f: A => B) =
        _ map f
    }

}

File: src/L03/Fuunctor.hs
module L03.Fuunctor where

import L01.Id
import L01.Optional
import L01.Validation
import L02.List

class Fuunctor f where
  fmaap :: (a -> b) -> f a -> f b

-- Exercise 1
-- Relative Difficulty: 1
--
-- | Maps a function on the Id functor.
--
-- Examples:
--
-- >>> fmaap (+1) (Id 2)
-- Id 3
instance Fuunctor Id where
  fmaap =
    error "todo"

-- Exercise 2
-- Relative Difficulty: 2
--
-- | Maps a function on the List functor.
--
-- Examples:
--
-- >>> fmaap (+1) Nil
-- []
--
-- >>> fmaap (+1) (1 :. 2 :. 3 :. Nil)
-- [2,3,4]
instance Fuunctor List where
  fmaap =
    error "todo"

-- Exercise 3
-- Relative Difficulty: 2
--
-- | Maps a function on the Optional functor.
--
-- Examples:
--
-- >>> fmaap (+1) Empty
-- Empty
--
-- >>> fmaap (+1) (Full 2)
-- Full 3
instance Fuunctor Optional where
  fmaap =
    error "todo"

-- Exercise 4
-- Relative Difficulty: 3
--
-- | Maps a function on the reader ((->) t) functor.
--
-- Examples:
--
-- >>> fmaap (+1) (*2) 8
-- 17
instance Fuunctor ((->) t) where
  fmaap =
    error "todo"

-- Exercise 5
-- Relative Difficulty: 2
instance Fuunctor IO where
  fmaap =
    error "todo"

-----------------------
-- SUPPORT LIBRARIES --
-----------------------

instance Fuunctor [] where
  fmaap = fmap
